[FEA] Generalized Adjustment Criterion

nparent1 opened the pull request:
Relates to feature https://github.com/py-why/dowhy/issues/402

This PR adds support for identifying generalized (non-backdoor) adjustment sets. Specifically, it adds support for finding a  minimal adjustment set if one exists (it is guaranteed to find a set if one does exist). Ongoing work in the pywhy-graphs library to enumerate all m-separating sets in causal graphs will later unlock the ability to enumerate all generalized adjustment sets.

The PR roughly adds the following
* A new field to the IdentifiedEstimand class to store general adjustment sets (given how ubiquitous the backdoor criterion is I left the backdoor fields alone and created new fields for these general sets)
* Support in the auto_identifier.py file's `identify_ate_effect()` method to compute a general adjustment set to return in an IdentifiedEstimand  object
   * Finding the adjustment set is all handled in a new `identify_generalized_adjustment_set()` method
* Refactoring to share various utility functions between the backdoor adjustment sets and the general adjustment sets 
* Unit tests to test a few cases of the generalized adjustment criterion (specifically cases where the backdoor criterion is not met)

In a subsequent PR I can add support for these general adjustment sets in the causal estimation stage.

References:
------------------------
Benito van der Zander, Maciej Liśkiewicz, and Johannes Textor. "Constructing Separators and
       Adjustment Sets in Ancestral Graphs." In Proceedings of UAI 2014, pages 907–916,
       2014.

nparent1 commented on dowhy/causal_identifier/adjustment_set.py:
```diff
@@ -0,0 +1,27 @@
+class AdjustmentSet:
```
Currently, dowhy stores adjustment sets as a list of dictionaries where the dictionaries will typically look like {"backdoor_set": [...]} (but will occasionally have an additional "num_paths_blocked_by_observed_nodes " field). This means in several places across the repo the raw string "backdoor_set" has to be used to access the actual adjustment set. 

I've added this class so that one can more safely access the adjustment set (eg by calling the getter method) rather than using the raw string, and also so that we can support and delineate new adjustment sets (eg the generalized adjustment set, which generalizes the backdoor criterion)

nparent1 commented on dowhy/graph.py:
```diff
@@ -187,13 +188,54 @@ def is_blocked(graph: nx.DiGraph, path, conditioned_nodes):
         return False
 
 
+def get_ancestors(graph: nx.DiGraph, nodes):
```
I noticed that get_descendants (line 198) does not include the nodes themselves, so I've replicated that here. Though when we invoke this method below we will have to union the result with the nodes themselves.

nparent1 commented:
This PR needs to bump the networkx dependency since the  algorithm wasnt added until [release 3.0](https://github.com/networkx/networkx/releases/tag/networkx-3.0rc1) (see [this PR](https://github.com/networkx/networkx/pull/5898))

nparent1 commented on dowhy/causal_identifier/auto_identifier.py:
```diff
@@ -290,6 +308,32 @@ def identify_ate_effect(
+    # This generalizes the backdoor criterion, identifying other valid covariate adjustment sets that might not
+    # satisfy the backdoor criterion. This capability requires python >=3.10
+    adjustment_variables_dict = default_adjustment_id = None
+    if sys.version_info >= (3, 10):
```
Only identify the generalized adjustment criterion if the user has python version >=3.10, so we can use networkx's find_minimal_d_separator algorithm (which requires networkx >=3.3)

nparent1 commented:
Converting to a draft since it seems some tests are failing, I will reopen once I resolve

nparent1 commented on pyproject.toml:
```diff
@@ -71,7 +71,10 @@ pandas = [
     {version = ">1.0", python = ">=3.9"}
 ]
-networkx = ">=2.8.5"
+networkx = [
```
needs to bump the networkx dependency since the find_minimal_d_separator  algorithm wasnt available in release 2.8.5 (see https://github.com/networkx/networkx/pull/5898) - this requires python>=3.10, which means our generalized adjustment will only be available to users with python >=3.10

amit-sharma replied:
That should be okay. This edit makes sense.

amit-sharma reviewed (commented):
Thanks for starting this @nparent1 
Two requests as I review this PR:
1. Can you add a notebook showing how the new feature can be used? you can add a notebook under `docs/source/example_notebooks/`. As examples, you can refer to this notebook on [ID algorithm](https://www.pywhy.org/dowhy/v0.12/example_notebooks/identifying_effects_using_id_algorithm.html) or the [optimal adjustment criterion](https://www.pywhy.org/dowhy/v0.12/example_notebooks/dowhy_efficient_backdoor_example.html). It will be ideal if the notebook can contrast with the output of the backdoor criterion and show when the generalized adjustment criterion is useful. 
2. Can you add documentation to all the user-facing functions? You can refer to auto-identifier.py or other files. I would suggest adding `:param:` and `:returns:` documentation for the functions at the minimum.

nparent1 commented:
Done! @amit-sharma

I've added an example notebook and documentation.

The example notebook is at: docs/source/example_notebooks/dowhy_generalized_covariate_adjustment_example.ipynb

The only user-facing function I had envisioned at this point was the `identify_effect_auto()` method in auto_identifier.py (and CausalModel's `identify_effect()` method, but my changes don't touch its method header). I also added documentation to my two utility functions in the graph.py file though, since I saw some of those methods have detailed documentation.

One other flag - in putting together the notebook I think I found a small bug in the frontdoor criterion codepath, which I've addressed in this PR since it prevented me from running  identify_effect() on my example graph (I'll add a comment to the PR pointing out the change so you can verify that it is appropriate)

nparent1 commented on dowhy/causal_identifier/auto_identifier.py:
```diff
@@ -795,7 +847,10 @@ def identify_frontdoor(
     eligible_variables = (
-        get_descendants(graph, action_nodes) - set(outcome_nodes) - set(get_descendants(graph, outcome_nodes))
+        get_descendants(graph, action_nodes)
+        - set(action_nodes)
```
I added this line to remove the action nodes from the set of eligible nodes (and added a unit test which illustrates why this is needed). If there are multiple action nodes and one of them ends up in the list of eligible variables, we can run into a networkx error when calling the is_d_separator method, in which we are checking if one of the action nodes is a d-separator for the set of action nodes

amit-sharma replied:
ah, good catch. Yes, this makes sense, thanks for adding it.

nparent1 commented on tests/causal_identifiers/example_graphs.py:
```diff
@@ -422,4 +483,15 @@
         valid_frontdoor_sets=[],
         invalid_frontdoor_sets=[{"Z"}, {"M1"}, {"M2"}, {"M1", "M2"}],
     ),
+    # This example is reproduced from the generalized_adjustment examples, and is
```
Here is the unit test I've added to illustrate why we need to filter out the action nodes from the set of eligible notes for the frontdoor criterion (without my change this test throws an error)

amit-sharma commented on dowhy/causal_identifier/auto_identifier.py:
```diff
@@ -158,6 +174,9 @@ def identify_effect_auto(
     :param conditional_node_names: variables that are used to determine treatment. If none are
     provided, it is assumed that the intervention is static.
+    :param generalized_adjustment: specify whether to return a single minimal adjustment set which
+    matches the general adjustment criterion ("default"), or two exhaustively compute all such adjustment sets ("exhaustive-search"). For now
```
typo: two exhaustively -> to exhaustively?

amit-sharma commented on dowhy/causal_identifier/auto_identifier.py:
```diff
@@ -844,6 +899,48 @@ def identify_frontdoor(
+            return []
+        return [AdjustmentSet(AdjustmentSet.GENERAL, adjustment_set)]
+    elif generalized_adjustment == GeneralizedAdjustment.GENERALIZED_ADJUSTMENT_EXHAUSTIVE:
+        raise ValueError("Exhaustive identification of general adjustment sets is not yet supported.")
```
How about adding a final "else" clause that raises a ValueError: "Please provide a valid type of Generalized Adjustment". May make the code clearer to follow.

amit-sharma commented on tests/causal_identifiers/example_graphs.py:
```diff
@@ -422,4 +483,15 @@
     ),
+    # This example is reproduced from the generalized_adjustment examples, and is
+    # added to test that the frontdoor criterion successfully filters out all the action
+    # nodes as ineligible variables for the
```
"for the"---perhaps delete?

amit-sharma reviewed (commented):
Thanks for adding this PR, @nparent1. Really appreciate the comprehensive test cases and the enhanced docstrings. 

I think this is ready to be merged. I just flagged a few typos. once you address them, I'm happy to merge.

nparent1 commented:
Thanks for reviewing @amit-sharma ! Final typos have been fixed
